File: data/Demon_matlab/DemonAnalysis3.m
function[sd, td, fd] = DemonAnalysis2(audio_data, Fs, pars)

% bandpass filter

% Design the filter
[b, a] = butter(pars.DEMON.bandpass.order, [pars.DEMON.bandpass.f_low/(Fs/2), .99], 'bandpass');

% Apply the filter to your audio data (assuming it's stored in a variable called 'audio_data')
filtered_audio = filter(b, a, audio_data);

% Define the length of each window (td)
d = round(1/pars.DEMON.fsd * Fs); % You can adjust this value as needed
fsd1 = Fs/d;

% Calculate the number of windows
num_windows = floor(length(filtered_audio) / d);

% Initialize an array to store RMS values for each part
rms_values = zeros(1, num_windows);

% Perform Hilbert transform and calculate RMS for each part
clear rms_values
for i = 1:num_windows

    % Define the start and end indices of the current window
    start_idx = (i - 1) * d + 1;
    end_idx = min(i * d, length(filtered_audio));

    % Extract the current window
    window_data = filtered_audio(start_idx:end_idx);

    % Mark Trevorrows suggestion for finding the envelope:
    rms_values(i) = sqrt(mean(abs(hilbert(window_data)).^2 + abs(window_data).^2));

end

% Spectrogram
df1 = 1/(pars.spectrogram.N_d/fsd1);
[sd, fd, td] = spectrogram(rms_values, pars.spectrogram.N_d, 5*pars.spectrogram.N_d/6, pars.DEMON.minfreq: df1 :pars.DEMON.maxfreq, fsd1);

return

File: Sound_data/Demon_analysis/DemonAnalysis3.py
import numpy as np
import scipy.signal as signal
from scipy.signal import butter, hilbert, spectrogram

def demon_analysis(audio_data, Fs, pars):
    # Design the bandpass filter
    b, a = butter(pars['DEMON']['bandpass']['order'], 
                  [pars['DEMON']['bandpass']['f_low'] / (Fs / 2), 0.99], 
                  btype='bandpass')

    # Apply the filter
    filtered_audio = signal.filtfilt(b, a, audio_data)

    # Define window length
    d = round(1 / pars['DEMON']['fsd'] * Fs)
    fsd1 = Fs / d

    # Calculate number of windows
    num_windows = len(filtered_audio) // d

    # RMS calculation
    rms_values = np.zeros(num_windows)
    for i in range(num_windows):
        start_idx = i * d
        end_idx = min((i + 1) * d, len(filtered_audio))
        window_data = filtered_audio[start_idx:end_idx]

        # Envelope detection using Hilbert transform
        rms_values[i] = np.sqrt(np.mean(np.abs(hilbert(window_data))**2 + np.abs(window_data)**2))

    # Spectrogram
    df1 = 1 / (pars['spectrogram']['N_d'] / fsd1)
    print(f"nfft = {int((pars['DEMON']['maxfreq'] - pars['DEMON']['minfreq']) / df1)}")
    print(f"nperseg = {pars['spectrogram']['N_d']}")
    f, t, sd = spectrogram(rms_values, fsd1, 
                           nperseg=pars['spectrogram']['N_d'], 
                           noverlap=5 * pars['spectrogram']['N_d'] // 6,
                           nfft=int((pars['DEMON']['maxfreq'] - pars['DEMON']['minfreq']) / df1),
                           scaling='density')

    return sd, t, f
